Sources/Defaults/Utilities.swift

```diff
@@ -230,6 +230,179 @@ extension Defaults.Serializable {
 	}
 }
 
+/**
+A reader/writer threading lock based on `libpthread`.
+*/
+final class RWLock {
+	private let lock: UnsafeMutablePointer<pthread_rwlock_t> = UnsafeMutablePointer.allocate(capacity: 1)
+
+	init() {
+		let err = pthread_rwlock_init(lock, nil)
+		precondition(err == 0, "\(#function) failed in pthread_rwlock_init with error \(err)")
+	}
+
+	deinit {
+		let err = pthread_rwlock_destroy(lock)
+		precondition(err == 0, "\(#function) failed in pthread_rwlock_destroy with error \(err)")
+		lock.deallocate()
+	}
+
+	private func lockRead() {
+		let err = pthread_rwlock_rdlock(lock)
+		precondition(err == 0, "\(#function) failed in pthread_rwlock_rdlock with error \(err)")
+	}
+
+	private func lockWrite() {
+		let err = pthread_rwlock_wrlock(lock)
+		precondition(err == 0, "\(#function) failed in pthread_rwlock_wrlock with error \(err)")
+	}
+
+	private func unlock() {
+		let err = pthread_rwlock_unlock(lock)
+		precondition(err == 0, "\(#function) failed in pthread_rwlock_unlock with error \(err)")
+	}
+
+	@inlinable
+	func withReadLock<R>(body: () -> R) -> R {
+		lockRead()
+		defer {
+			unlock()
+		}
+		return body()
+	}
+
+	@inlinable
+	func withWriteLock<R>(body: () -> R) -> R {
+		lockWrite()
+		defer {
+			unlock()
+		}
+		return body()
+	}
+}
+
+/**
+A queue for executing asynchronous tasks in order.
+
+```swift
+actor Counter {
+	var count = 0
+
+	func increase() {
+		count += 1
+	}
+}
+let counter = Counter()
+let queue = TaskQueue(priority: .background)
+queue.async {
+	print(await counter.count) //=> 0
+}
+queue.async {
+	await counter.increase()
+}
+queue.async {
+	print(await counter.count) //=> 1
+}
+```
+*/
+final class TaskQueue {
+	typealias AsyncTask = @Sendable () async -> Void
+	private var queueContinuation: AsyncStream<AsyncTask>.Continuation?
+
+	init(priority: TaskPriority? = nil) {
+		let taskStream = AsyncStream<AsyncTask> { queueContinuation = $0 }
+
+		Task.detached(priority: priority) {
+			for await task in taskStream {
+				await task()
+			}
+		}
+	}
+
+	deinit {
+		queueContinuation?.finish()
+	}
+
+	/**
+	Queue a new asynchronous task.
+	*/
+	func async(_ task: @escaping AsyncTask) {
+		queueContinuation?.yield(task)
+	}
+
+	/**
+	Queue a new asynchronous task and wait until it done.
+	*/
+	func sync(_ task: @escaping AsyncTask) {
+		let semaphore = DispatchSemaphore(value: 0)
+
+		queueContinuation?.yield {
+			await task()
+			semaphore.signal()
+		}
+
+		semaphore.wait()
```

Comment:
This will deadlock if the `task` accidentally calls something that also calls `.sync()` (nested calls).

According to ChatGPT, you could do something like this:

```swift
final class TaskQueue {
	typealias AsyncTask = @Sendable () async -> Void
	private var queueContinuation: AsyncStream<AsyncTask>.Continuation?
	private let queue: DispatchQueue

	init(priority: TaskPriority? = nil) {
		queue = DispatchQueue(label: "com.example.taskqueue", attributes: .concurrent)
		let taskStream = AsyncStream<AsyncTask> { queueContinuation = $0 }

		Task.detached(priority: priority) {
			for await task in taskStream {
				await task()
			}
		}
	}

	deinit {
		queueContinuation?.finish()
	}

	func async(_ task: @escaping AsyncTask) {
		queue.async {
			self.queueContinuation?.yield(task)
		}
	}

	func sync(_ task: @escaping AsyncTask) {
		let semaphore = DispatchSemaphore(value: 0)

		queue.async {
			if DispatchQueue.getSpecific(key: self.queue.key) == self.queue.id {
				Task {
					await task()
					semaphore.signal()
				}
			} else {
				self.queueContinuation?.yield {
					await task()
					semaphore.signal()
				}
			}
		}

		semaphore.wait()
	}

	func flush() async {
		await withCheckedContinuation { continuation in
			queue.async {
				self.queueContinuation?.yield {
					continuation.resume()
				}
			}
		}
	}
}
```

Reply:
That's true, I think we should deprecate `sync`. 
Would it be acceptable to use `TaskQueue.async` for updating remote key changes?

Reply:
The async queue is concurrent, so it maybe cause ordering problems.

Reply:
`TaskQueue.async` is not concurrent; it will run similarly to `DispatchQueue.async`. 
I believe it should not have ordering problems.

Reply:
Btw, if calling `await flush` in `TaskQueue.async`, it will also cause deadlock.
Maybe we should emit a Xcode runtime warning to prevent user doing this.

ex. 

```swift
let queue = TaskQueue(priority: .background)
queue.async {
    print("1")
    queue.async {
        print("2")
    }
    await queue.flush()
}
await queue.flush()
// => 1 "2" will never print.
```

Reply:
> Btw, if calling await flush in TaskQueue.async, it will also cause deadlock.
Maybe we should emit a Xcode runtime warning to prevent user doing this.

Users don't have access to `TaskQueue` though. It's an internal type.